chainer/training/trainer.py

```diff
@@ -0,0 +1,285 @@
+        self.updater = updater
+        self.stop_trigger = trigger_module.get_trigger(stop_trigger)
+        self.observation = {}
+        self.out = out
+
+        reporter = reporter_module.Reporter()
+        for name, optimizer in six.iteritems(updater.get_all_optimizers()):
+            reporter.add_observer(name, optimizer.target)
+            reporter.add_observers(
+                name, optimizer.target.namedlinks(skipself=True))
+        self.reporter = reporter
+
+        self._done = False
+        self._extensions = collections.OrderedDict()
+
+        updater.connect_trainer(self)
+
+    def extend(self, extension, name=None, trigger=None, priority=None,
+               invoke_before_training=None):
+        """Registers an extension to the trainer.
+
+        :class:`Extension` is a callable object which is called after each
+        update unless the corresponding trigger object decides to skip the
+        iteration. The order of execution is determined by priorities:
+        extensions with higher priorities are called earlier in each iteration.
+        Extensions with the same priority are invoked in the order of
+        registrations.
+
+        If two or more extensions with the same name are registered, suffixes
+        are added to the names of the second to last extensions. The suffix is
+        ``_N`` where N is the ordinal of the extensions.
+
+        See :class:`Extension` for the interface of extensions.
+
+        Args:
+            extension: Extension to register.
+            trigger (tuple or Trigger): Trigger object that determines when to
+                invoke the extension. If it is None, ``extension.trigger`` is
+                used instead. If the trigger is not callable, it is passed to
+                :class:`IntervalTrigger` to build an interval trigger.
+            priority (int): Invocation priority of the extension. Extensions
+                are invoked in the descending order of priorities in each
+                iteration. If this is None, ``extension.priority`` is used
+                instead.
+            invoke_before_training (bool): If True, the extension is also
+                invoked just before entering the training loop. If this
+                None, ``extension.invoke_before_training`` is used instead.
+                This option is mainly used for extensions that alter the
+                training configuration (e.g., learning rates); in such a case,
+                resuming from snapshots require the call of extension to
+                recover the configuration before any updates.
+
+        """
+        if name is None:
+            name = getattr(extension, 'name', None)
+            if name is None:
+                name = getattr(extension, 'default_name', None)
+                if name is None:
+                    raise TypeError('name is not given for the extension')
+                elif name == 'training':
```

Comment:
If we feed `training` with `name` argument of this method, this check is not invoked. Is it acceptable?